In [ ]:
%matplotlib inline
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np
import pandas as pd
from IPython.display import display
from IPython.display import Image
sns.set_context("notebook")

In [55]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


# Extracting Data and Filtering
:The `[]` operator supports array of booleans as an input. In this case, the array must be exactly as long as the number of rows.

The result is a **filtered version of the data frame**, where **only rows corresponding to True appear**.

# Q1.
Select multiple criteria by creating multiple boolean Series and combining them using the `&` operator. For Examole Result is win and %win is more than 50.

In [ ]:
 # reading a CSV file into a data frame

df = pd.read_csv("/content/elections.csv")
df

 #Note: if we end a cell with an expression or variable name, the result will print the data frame



Year          Candidate                  Party  Popular vote Result  \
0    1824     Andrew Jackson  Democratic-Republican        151271   loss   
1    1824  John Quincy Adams  Democratic-Republican        113142    win   
2    1828     Andrew Jackson             Democratic        642806    win   
3    1828  John Quincy Adams    National Republican        500897   loss   
4    1832     Andrew Jackson             Democratic        702735    win   
..    ...                ...                    ...           ...    ...   
177  2016         Jill Stein                  Green       1457226   loss   
178  2020       Joseph Biden             Democratic      81268924    win   
179  2020       Donald Trump             Republican      74216154   loss   
180  2020       Jo Jorgensen            Libertarian       1865724   loss   
181  2020     Howard Hawkins                  Green        405035   loss   

             %  
0    57.210122  
1    42.789878  
2    56.203927  
3    43.796073  
4    54.574789  
..         ...  
177   1.073699  
178  51.311515  
179  46.858542  
180   1.177979  
181   0.255731  

[182 rows x 6 columns]

In [ ]:
#apply a boolean condition using the appropriate operator on 2 different conditions like Result should be WIN and % should be GREATER THAN 50

We will nowe work with a dataset containing the names of all babies born in California, USA. Visit the [Social Security Adminisitration Website](https://www.ssa.gov/oact/babynames/limits.html) for more details about the dataset if you want.

**Let's start by loading the Baby Names dataset.**

In [ ]:
import urllib.request
import os.path
import zipfile
import pandas as pd

data_url = "https://www.ssa.gov/oact/babynames/state/namesbystate.zip"
local_filename = "C:\\Users\\Admin\\Downloads\\babynamesbystate.zip"
if not os.path.exists(local_filename): # if the data exists don't download again
    with urllib.request.urlopen(data_url) as resp, open(local_filename, 'wb') as f:
        f.write(resp.read())

zf = zipfile.ZipFile(local_filename, 'r')

ca_name = 'CA.TXT'
field_names = ['State', 'Sex', 'Year', 'Name', 'Count']
with zf.open(ca_name) as fh:
    babynames = pd.read_csv(fh, header=None, names=field_names)

babynames.head(15)

State Sex  Year       Name  Count
0     CA   F  1910       Mary    295
1     CA   F  1910      Helen    239
2     CA   F  1910    Dorothy    220
3     CA   F  1910   Margaret    163
4     CA   F  1910    Frances    134
5     CA   F  1910       Ruth    128
6     CA   F  1910     Evelyn    126
7     CA   F  1910      Alice    118
8     CA   F  1910   Virginia    101
9     CA   F  1910  Elizabeth     93
10    CA   F  1910   Florence     93
11    CA   F  1910      Marie     90
12    CA   F  1910    Mildred     90
13    CA   F  1910       Rose     74
14    CA   F  1910      Hazel     68

In [ ]:
#dsplay size

baby_name_size =babynames.size
baby_name_size

2069470

# Q2.
Calculate the some statics of babynames DataFrame using Utility Functions, like .describe(), .size, .shape and report your analysis in comments

In [ ]:
babynames.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 413894 entries, 0 to 413893
Data columns (total 5 columns):
 #   Column  Non-Null Count   Dtype 
---  ------  --------------   ----- 
 0   State   413894 non-null  object
 1   Sex     413894 non-null  object
 2   Year    413894 non-null  int64 
 3   Name    413894 non-null  object
 4   Count   413894 non-null  int64 
dtypes: int64(2), object(3)
memory usage: 15.8+ MB


In [ ]:
# code here
babynames.shape

(413894, 5)

In [ ]:
#code here
babynames.describe()

Year          Count
count  413894.000000  413894.000000
mean     1986.314353      79.140348
std        27.190404     292.028292
min      1910.000000       5.000000
25%      1969.000000       7.000000
50%      1992.000000      13.000000
75%      2008.000000      38.000000
max      2023.000000    8263.000000

In [ ]:
#code here
babynames.size

2069470

**My Obeservation**



*   The total coloumns in this data is **5**. The **3**  columns is **object** data type and **2** is **int64**
*   There is **0 null values** in this data.


*   We get information from two selected coloumns **year** & **Count**.
We get Total count **413894** which both columns.

*   We get **Mean** of **Year** & **Count** which is **1986.31** & **79.14** respectivily.


*   The **Min** of year is **1910** and **Max** is **2023**.
same as with count the **Max** value is **8263** & **Min** value is **5**



*   The **25%**,**50%** and **75%** percentile of year is **1969**,**1992** and **2008** respectivily.






# Q.3:

Using the babynames DataFrame that contains baby names data for the state of California, how can you use the 'isin' method to find out which records have names that are in a given list of names, such as "Emma," "Liam," and "Olivia," and what would be the output of this operation when applied to the first 15 rows of the DataFrame (babynames.head(15))?

In [ ]:
babynames.head()

State Sex  Year      Name  Count
0    CA   F  1910      Mary    295
1    CA   F  1910     Helen    239
2    CA   F  1910   Dorothy    220
3    CA   F  1910  Margaret    163
4    CA   F  1910   Frances    134

In [ ]:
# List of names to check

list_name = ["Emma", "Liam","Olivia"]
# Use the isin method to filter records with names in the name_list
babynames[babynames["Name"].isin(list_name)]
# Display the filtered DataFrame with appropriate head value i.e., 15
babynames.head(15)

State Sex  Year       Name  Count
0     CA   F  1910       Mary    295
1     CA   F  1910      Helen    239
2     CA   F  1910    Dorothy    220
3     CA   F  1910   Margaret    163
4     CA   F  1910    Frances    134
5     CA   F  1910       Ruth    128
6     CA   F  1910     Evelyn    126
7     CA   F  1910      Alice    118
8     CA   F  1910   Virginia    101
9     CA   F  1910  Elizabeth     93
10    CA   F  1910   Florence     93
11    CA   F  1910      Marie     90
12    CA   F  1910    Mildred     90
13    CA   F  1910       Rose     74
14    CA   F  1910      Hazel     68

# Q.4:
How can you use the 'str.startswith' method in pandas to filter records in a babyname DataFrame? Provide an example code using the babynames DataFrame and explain how this method can be employed to extract data for baby names that begin with 'A' or 'B' set of letters.

In [ ]:
# Filter rows where the 'Name' column starts with 'A' OR 'B'

babynames_A_B = babynames[(babynames["Name"].str.startswith("A")) | (babynames["Name"].str.startswith("B"))]

# Display the filtered DataFrame
print("Baby names starting with 'A' or 'B':")
print(babynames_A_B.head(5))  # Display the first 5 names starting with 'A' or 'B'


Baby names starting with 'A' or 'B':
   State Sex  Year     Name  Count
7     CA   F  1910    Alice    118
23    CA   F  1910  Bernice     59
31    CA   F  1910     Anna     51
32    CA   F  1910      Ann     47
36    CA   F  1910    Betty     40


# Q.5

Find the most popular baby name in California in 2012

__Exercise__: Try to find the most popular baby name by finding the maximum value for 'Count' and then apply filtering to find the name.


In [ ]:
from os import popen
# Step 5: Find the maximum count and the corresponding name(s)
# Filter the dataset for the year 2012
babynames_2012 = babynames[babynames['Year'] == 2012]

# Find the row with the maximum 'Count' in the filtered dataset
most_popular_name_2012 = babynames_2012[babynames_2012['Count'] == babynames_2012['Count'].max()]
# Display the most popular name and its count
print(most_popular_name_2012[['Name','Year','Count']])



#Hint: You should apply max on the count but try to get the whole data of that maximum row rather than try to get only the max count.


          Name  Year  Count
197543  Sophia  2012   3645


# Q. 6
Find baby names that start with J, in th Year from 2001 to 2019

*Note: To answer this question, we need a way to look at the individual characters of the baby names*

In [ ]:
# Note: Series has powerful string manipulation methods associated with it

# Step 5: Find names that start with 'J'

# Filter rows for names starting with 'J' and for the years between 2001 and 2019
babynames_J_2001_2019 = babynames[(babynames["Name"].str.startswith("J")) &  (babynames["Year"].between(2001, 2019))]

# Display the filtered DataFrame
print("Baby names starting with 'J' from 2001 to 2019:")


#Hint: 3 different conditions in a single boolean condition can be used using an operator

# Step 6: Display the filtered DataFrame with first 15 rows
babynames_J_2001_2019.head(15)

Baby names starting with 'J' from 2001 to 2019:


State Sex  Year        Name  Count
152833    CA   F  2001     Jessica   2244
152835    CA   F  2001    Jennifer   2027
152839    CA   F  2001     Jasmine   1655
152854    CA   F  2001  Jacqueline   1195
152864    CA   F  2001       Julia    983
152873    CA   F  2001     Jocelyn    856
152920    CA   F  2001      Jordan    510
152925    CA   F  2001      Jasmin    500
152929    CA   F  2001     Juliana    494
152931    CA   F  2001       Jenna    490
152938    CA   F  2001      Jazmin    465
152952    CA   F  2001      Joanna    396
152955    CA   F  2001        Jade    384
152969    CA   F  2001     Jazmine    346
152997    CA   F  2001    Julianna    282

# Q.7
Add a new column called "name_length" to the babynames DataFrame, you can use .str.len() in existing Name column. Moreover. also modify the Length values with new values like existing_value-1 and rename the length column with new label as desceribe in slides.



In [ ]:
# Calculate the length of each name and add it as a new column "_length"

name_length = babynames["Name"].str.len()

babynames["_length"] = name_length

# Modify 'name_length' values to be existing_value - 1
babynames["_length"] = name_length-1
babynames

# Rename the 'name_length' column to the new label 'name_length_modified'

babynames["name_length_modified"] = name_length
babynames


State Sex  Year      Name  Count  _length  name_length_modified
0         CA   F  1910      Mary    295        3                     4
1         CA   F  1910     Helen    239        4                     5
2         CA   F  1910   Dorothy    220        6                     7
3         CA   F  1910  Margaret    163        7                     8
4         CA   F  1910   Frances    134        6                     7
...      ...  ..   ...       ...    ...      ...                   ...
413889    CA   M  2023      Ziah      5        3                     4
413890    CA   M  2023    Ziaire      5        5                     6
413891    CA   M  2023    Zidane      5        5                     6
413892    CA   M  2023      Zyan      5        3                     4
413893    CA   M  2023      Zyir      5        3                     4

[413894 rows x 7 columns]

# Q.8

Use .str.split() to sample the names with 'ad' string and show first 10 samples.

In [ ]:
# Step 4: Filter names that contain 'ad' using str.contains and store it in th variable named ad_names

ad_name = babynames[babynames["Name"].str.contains("ad")]
ad_name

# Step 5: Use .str.split() to split names containing 'ad' in ad_names (for demonstration purposes)
names_with_ad =babynames["Name"].str.split("ad")
names_with_ad

0             [Mary]
1            [Helen]
2          [Dorothy]
3         [Margaret]
4          [Frances]
             ...    
413889        [Ziah]
413890      [Ziaire]
413891      [Zidane]
413892        [Zyan]
413893        [Zyir]
Name: Name, Length: 413894, dtype: object

More on str methods when we'll study regular expressions in this course. Note that there are other approaches that we can use to find the same answer as above.

# Q.9
 Sort names by their length using lambda function.

In [ ]:
babynames.columns

Index(['State', 'Sex', 'Year', 'Name', 'Count'], dtype='object')

In [ ]:
#Hint: focus on 2 columns here i.e., NAME and LENGTH (apply appropriate parameters using lambda)

# Create the Name_Length column
babynames["Name_Length"] = babynames["Name"].apply(lambda x: len(x))

# Sort by both 'Name' and 'Name_Length' using different lambda functions
babynames.sort_values(
    by=["Name", "Name_Length"],
    key=lambda x: x.str.len() if x.name == "Name" else x,
    ascending=[True, False],
    inplace=True
)
babynames

State Sex  Year             Name  Count  name_length_modified  \
4454      CA   F  1920               Jo      5                     2   
4908      CA   F  1921               Jo      9                     2   
6157      CA   F  1923               Jo      8                     2   
6671      CA   F  1924               Jo     16                     2   
7350      CA   F  1925               Jo     17                     2   
...      ...  ..   ...              ...    ...                   ...   
337819    CA   M  1996  Franciscojavier      8                    15   
340954    CA   M  1997  Franciscojavier      5                    15   
341130    CA   M  1997  Ryanchristopher      5                    15   
343125    CA   M  1998  Franciscojavier      6                    15   
348093    CA   M  2000  Franciscojavier      6                    15   

        Name_Length  
4454              2  
4908              2  
6157              2  
6671              2  
7350              2  
...             ...  
337819           15  
340954           15  
341130           15  
343125           15  
348093           15  

[413894 rows x 7 columns]

# Q.10
1: The code randomly mixes up the list of baby names. Why might someone want to do this, and how could it affect the analysis of the names?

2: After mixing, the code figures out how long each baby name is. Why would we care about name lengths, and what can this tell us about the names?

3: The code arranges the name lengths from longest to shortest. Why might we want to do this, and what could it help us find in the data?

4: The code finds the order of baby names based on their lengths and shows us the first few names in this new order. When might we need to organize names this way, and what might we discover from the first few names in this order?

In [45]:
#This is an entirely theoretical question and you are required to explain your logic in your own words. Make sure not to plagarise your content.

#Part1, shuffling can be done using sample(frac=1)

It affect the actual Data to decrease  biasness and increased randomness  for machine learning task,
but it can hide important pattern trend over time.

# #Part2

Looking at baby name lengths helps us see trends, like whether short or long names are more popular, and can tell us about cultural preferences.

# #Part3
Sorting name lengths from longest to shortest can help us spot trends in popularity and see if longer names are favored in certain cultures or groups.

# #Part4
Organizing names by length helps identify trends, and the first few names can reveal popular styles or preferences for longer names in certain demographics.


# Bonus Question

1: How can you use the .sample() method to randomly select a single row from the babynames DataFrame, and what is the purpose of such a selection?

2: Suppose you want to generate a random sample of 10 baby names from the babynames DataFrame. How would you use the .sample() method to achieve this, and what parameters or arguments would you specify?

3: You want to perform a bootstrap resampling analysis on the baby names data in the babynames DataFrame. Describe how you can repeatedly sample with replacement using .sample() to create multiple bootstrap samples for statistical analysis

In [54]:
#Part 1

#Use .sample() and explain why should/could we use it to display a single row

babynames["Name"].sample(20)
babynames

# Because we see randomly the length of names, we Deal only Name column to check quick overview about the entries.

#Part 2

#Use .sample() and explain why should/could we use it to display a ten rows.

babynames["Name"].sample(20).head(10)
# Because we want to quick overiew of entire entries for data integrity.

#Part 3

# Assuming you have the 'babynames' DataFrame

# Set the random seed for reproducibility (read reproducibility in documentation and check out what random_seed does)
import random

# Set the random seed
random.seed(42)  # You can use any number as the seed

# Or if using NumPy
import numpy as np

np.random.seed(42)

# Assuming babynames is your DataFrame
n_samples = 1000  # Number of bootstrap samples
sample_size = len(babynames)  # Sample size (same as the original dataset)

# Initialize lists to store results
means = []
medians = []

# Perform bootstrapping
for _ in range(n_samples):
    # Generate a bootstrap sample with replacement
    bootstrap_sample = babynames.sample(n=sample_size, replace=True)

    # Perform your statistical analysis on the bootstrap sample
    means.append(bootstrap_sample['Name_Length'].mean())  # Calculate mean
    medians.append(bootstrap_sample['Name_Length'].median())  # Calculate median

# Calculate overall statistics
mean_of_means = sum(means) / len(means)
median_of_medians = sum(medians) / len(medians)

# Output the results
print(f"Mean of Bootstrap Means: {mean_of_means}")
print(f"Mean of Bootstrap Medians: {median_of_medians}")



#Hint: The below link is fun to explore and figure out in more detail what all this is:
#https://stackoverflow.com/questions/31502958/drawing-a-bootstrap-sample-from-a-pandas-dataframe

Mean of Bootstrap Means: 5.95667440938984
Mean of Bootstrap Medians: 6.0
